### ISSUE TITLE
Need some way to check project is built with coverage properly (post build time).

### ISSUE THREAD

**Issue by inferno-chromium** [google/oss-fuzz#382]

**Comment by kcc** [comment]
One way to do this is to use `sancov -print-coverage-pcs $BINARY | wc -l` as the metric
(number of instrumented BBs). For the incorrect grpc build:
```
for f in /out/*; do printf "% 4d $f\n" $(sancov -print-coverage-pcs $f | wc -l); done 
 233 /out/api_fuzzer
  18 /out/client_fuzzer
   5 /out/fuzzer
   8 /out/fuzzer_response
   8 /out/fuzzer_serverlist
   7 /out/hpack_parser_fuzzer_test
   7 /out/percent_decode_fuzzer
   4 /out/percent_encode_fuzzer
   3 /out/request_fuzzer
   3 /out/response_fuzzer
  18 /out/server_fuzzer
  19 /out/ssl_server_fuzzer
   5 /out/uri_fuzzer_test
```

A threshold should probably be around 100-200 cov points. 
(It will fail to flag api_fuzzer, but will flag all others in grpc)

**Comment by oliverchang** [comment]
Could a better threshold be based on the number of functions we detect in the binary (via symbols)?

**Comment by kcc** [comment]
"number of functions" could be confusing. 
I've seen single-function fuzz targets with thousands of cov points (imagine a state machine).
You can also have a target with lots of non-instrumented functions linked statically from an external dep....

**Comment by MarkusTeufelberger** [comment]
If compiled with LTO/thinLTO, is it a reasonable expectation that all/most of the instrumentation points are actually reachable?

**Comment by kcc** [comment]
err... kind-of... There are lots of ways to fool LTO into not removing some essentially dead code.

**Comment by MarkusTeufelberger** [comment]
Hm, so at least it would be a way to measure the overall efficiency of LTO (how much was removed compared to a non-LTO build because LTO was sure that it is dead code), but not the overall efficiency of a fuzzer (how many % of all reachable instrumentation points were reached so far).

Maybe just tracking the number of instrumented points over time per fuzzer and alerting once it changes more than x% could be enough to spot errors/drastic changes. Might be better than magic values.

**Comment by inferno-chromium** [comment]
@MarkusTeufelberger - in this case, initial number is important too. When someone lands a new project, it is covering anything ? But you are right, we should continue to ensure that it remains at similar levels and does not fall off quickly (@tanin47 as fyi)

**Comment by kcc** [comment]
Yep, all that. LTO build for coverage analysis is in our plans. https://github.com/google/oss-fuzz/issues/191#issuecomment-267447359

**Comment by Dor1s** [comment]
Another "bad build" issue we've already seen several times (e.g. https://github.com/khaledhosny/oss-fuzz/commit/ad99a12a7c3a8649527d9425cda7d9847352623f or https://github.com/google/oss-fuzz/issues/568): incorrect memory tools used.

Not sure if that's a good solution, but we can check symbols to understand whether the build is correct or not, e.g.:
* bad ASan build:
```bash
$ objdump -dC ./ots-fuzzer | egrep "__ubsan|__msan" | wc -l
33188
```
* good UBSan build:
```bash
$ objdump -dC ../gdal/shape_fuzzer | egrep "__asan|__msan" | wc -l
0
```
* etc

**Comment by Dor1s** [comment]
Forgot to post notes from our last meeting:
* ASan and UBSan symbols can be presented in same binary, as UBSan usually shares ASan's runtime (though it has standalone runtime as well)
* ASan and MSan should not be presented in same binary, the approach above would be fine for ASan vs MSan check
* we may check for number of calls to UBsan runtime functions, if there are many such calls, most likely UBSan checks have been enabled.

Other than that, we may pass `-fno-sanitize=undefined`, but it would depend on how user appends other build flags, so 50/50 solution.

**Comment by Dor1s** [comment]
There are 3 issues to be addressed:

1) Missing or incorrect coverage instrumentation. Possible solutions:
  a) check number of instrumented BBs via sancov: https://github.com/google/oss-fuzz/issues/382#issuecomment-279807705
  b) rely on libFuzzer's error message: https://chromium.googlesource.com/chromium/llvm-project/llvm/lib/Fuzzer.git/+/cfd0bc33f4d449c4d41b16196b021c4975401478/FuzzerLoop.cpp#386

2) Multiple or incorrect sanitizers used. Possible solutions:
  a) https://github.com/google/oss-fuzz/issues/382#issuecomment-302038748 and https://github.com/google/oss-fuzz/issues/382#issuecomment-303662862

3) No sanitizer used. The solution most likely will be the same as for issue 2.

The solutions suggested above can be implemented in a separate script, let's say, `bad_build_check.sh` or something that should be invoked after compilation and before zipping stage.

Another suggested solution is to have a wrapper script for `clang` and `clang++`. The script should either verify compilation flags or record them for further verification. This might be a universal approach to target all the issues listed above, but also might be tricky for large projects.

**Comment by Dor1s** [comment]
Reminder for myself: look through the issues we've seen before https://github.com/google/oss-fuzz/pull/831

and then re-enable bad build checks: https://github.com/google/oss-fuzz/commit/718db9173d81905129fae15718aad65d9a8a9633#diff-b2d2c950972d10d3ddd6303fc7845502

**Comment by Dor1s** [comment]
Landed as #838
